File: frontend/lib/realtime/hubs/job-hub.ts
"use client";

import { useHubSubscription } from "@/lib/hooks/use-hub-subscription";

const PATH = "/hubs/jobs";

export type JobHubEvents = {
  TimelineUpdated?: (entry: unknown) => void;
  JobUpdated?: (job: unknown) => void;
  JobAdded?: (job: unknown) => void;
  JobRemoved?: (jobId: string) => void;
};

/** Jobs queue page — receives the cross-order queue stream. */
export function useJobQueueSubscription(events: JobHubEvents) {
  return useHubSubscription({
    hubPath: PATH,
    subscribeMethod: "SubscribeQueue",
    unsubscribeMethod: "UnsubscribeQueue",
    subscribeArgs: [],
    eventHandlers: events as Record<string, (...args: unknown[]) => void>,
  });
}

/** Job detail drawer — receives updates scoped to a single job. */
export function useJobSubscription(jobId: string | null, events: JobHubEvents) {
  return useHubSubscription({
    hubPath: PATH,
    subscribeMethod: "SubscribeJob",
    unsubscribeMethod: "UnsubscribeJob",
    subscribeArgs: [jobId],
    eventHandlers: events as Record<string, (...args: unknown[]) => void>,
    enabled: !!jobId,
  });
}

export const JOB_HUB_PATH = PATH;

File: src/DTMS.Api/Realtime/Hubs/JobHub.cs
using AMR.DeliveryPlanning.Api.Realtime.Hubs.Clients;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.SignalR;

namespace AMR.DeliveryPlanning.Api.Realtime.Hubs;

/// <summary>
/// Job-scoped realtime + the cross-order job queue. Two subscription
/// flavours so the Jobs queue page (broad) and a single Job drawer
/// (focused) can coexist without paying for each other's traffic.
/// </summary>
[Authorize]
public sealed class JobHub : Hub<IJobClient>
{
    private const string QueueGroup = "job-queue";

    /// <summary>
    /// Jobs queue page subscribes here to receive
    /// <see cref="IJobClient.JobUpdated"/> / <c>JobAdded</c> / <c>JobRemoved</c>
    /// pushed by the Job lifecycle projector.
    /// </summary>
    public Task SubscribeQueue()
        => Groups.AddToGroupAsync(Context.ConnectionId, QueueGroup);

    public Task UnsubscribeQueue()
        => Groups.RemoveFromGroupAsync(Context.ConnectionId, QueueGroup);

    public Task SubscribeJob(Guid jobId)
        => Groups.AddToGroupAsync(Context.ConnectionId, GroupKey(jobId));

    public Task UnsubscribeJob(Guid jobId)
        => Groups.RemoveFromGroupAsync(Context.ConnectionId, GroupKey(jobId));

    public static string GroupKey(Guid jobId) => $"job:{jobId:N}";
    public static string QueueGroupKey => QueueGroup;
}
